pkg/dataobj/internal/dataset/page_reader.go

```diff
@@ -0,0 +1,233 @@
+package dataset
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+)
+
+type pageReader struct {
+	page        Page
+	value       datasetmd.ValueType
+	compression datasetmd.CompressionType
+	ready       bool // Whether the pageReader is initialized for page.
+
+	lastValue    datasetmd.ValueType
+	lastEncoding datasetmd.EncodingType
+
+	closer      io.Closer
+	presenceDec *bitmapDecoder
+	valuesDec   valueDecoder
+
+	pageRow int64
+	nextRow int64
+}
+
+// newPageReader returns a new pageReader that reads from the provided page.
+// The page must hold values of the provided value type, and be compressed with
+// the provided compression type.
+func newPageReader(p Page, value datasetmd.ValueType, compression datasetmd.CompressionType) *pageReader {
+	var pr pageReader
+	pr.Reset(p, value, compression)
+	return &pr
+}
+
+// Read reads up to the next len(v) values from the page into v. It returns the
+// number of values read and any error encountered. At the end of the page,
+// Read returns 0, io.EOF.
+func (pr *pageReader) Read(ctx context.Context, v []Value) (n int, err error) {
+	// We need to initialize our readers before we can read from the page.
+	//
+	// If we've seeked backwards and our page row is now ahead of the row we want
+	// to read, we need to reinitialize the page reader to read from the start of
+	// the page.
+	if !pr.ready || pr.pageRow > pr.nextRow {
+		err := pr.init(ctx)
+		if err != nil {
+			return n, err
+		}
+	}
+
+	// "Skip" rows until we reach the starting row we want to read. We do this by
+	// reading garbage values into v.
+	for pr.pageRow < pr.nextRow {
```

Comment:
hopefully we can do better at some point to avoid decoding if possible

Reply:
That would require refactoring all the page decoders which is happening in #16418, but even a Seek there would still need this for loop to skip over values (but in those case it could discard them without any allocs). 

Definitely a future optimization we could look into!